File: problems/052/052.c
#include <math.h>
#include <stdio.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    const int NX    =  6;
    const int N_MAX = 10;

    int n           = 6;
    int num_1x      = 125874;

    int num_nx, same_digits;

    while (n != N_MAX)
    {
        const int NUM_MAX = (int)(5.0 * pow(10.0, (double)n - 1.0) / 3);

        while (++num_1x <= NUM_MAX)
        {
            num_nx = 0;
            char num_nx_str[NX][n];
            memset(&num_nx_str[0][0], 0, sizeof(num_nx_str));

            char digits_set_1[N_MAX+1];
            memset(&digits_set_1[0], '0', sizeof(digits_set_1));
            digits_set_1[N_MAX] = 0;

            for (int k = 0; k < NX; k++)
            {
                num_nx += num_1x;
                sprintf(&num_nx_str[k][0], "%d", num_nx);

                char digits_set_2[N_MAX+1];
                memset(&digits_set_2[0], '0', sizeof(digits_set_2));
                digits_set_2[N_MAX] = 0;

                for (int l = 0; l < NX; l++)
                    digits_set_2[num_nx_str[k][l] - '0'] = '1';

                if (k == 0)
                {
                    strcpy(&digits_set_1[0], &digits_set_2[0]);
                    continue;
                }

                same_digits = 1;

                if (strcmp(digits_set_1, digits_set_2) != 0)
                {
                    same_digits = 0;
                    break;
                }
            }

            if (same_digits)
                break;
        }

        if (same_digits)
            break;

        n++;
    }

    printf("%d\n", num_1x);

    return 0;
}

File: problems/052/052.f90
PROGRAM F052

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    INTEGER, PARAMETER :: NX    = 6
    INTEGER, PARAMETER :: N_MAX = 10

    CHARACTER (LEN=N_MAX), DIMENSION(NX) :: NUM_NX_STR
    CHARACTER (LEN=N_MAX)                :: DIGITS_SET_1, DIGITS_SET_2

    LOGICAL   :: SAME_DIGITS;

    INTEGER   :: N      = 6;
    INTEGER   :: NUM_1X = 125875;

    INTEGER   :: K, L, NUM_NX, NUM_MAX, TEMP_INDEX

    DO WHILE (N /= N_MAX)

        NUM_MAX = FLOOR(5.0 * (10.0 ** (N - 1)) / 3.0);

        DO WHILE (NUM_1X <= NUM_MAX)

            NUM_NX     = 0
            NUM_NX_STR = ACHAR(0)

            DIGITS_SET_1 = REPEAT("0", N_MAX)

            DO K = 1, NX

                NUM_NX = NUM_NX + NUM_1X
                WRITE(NUM_NX_STR(K)(1:N), '(I0)') NUM_NX

                DIGITS_SET_2 = REPEAT("0", N_MAX)

                DO L = 1, NX
                    TEMP_INDEX = IACHAR(NUM_NX_STR(K)(L:L)) - IACHAR('0') + 1
                    DIGITS_SET_2(TEMP_INDEX:TEMP_INDEX) = "1"
                END DO

                IF (K == 1) THEN
                    DIGITS_SET_1 = DIGITS_SET_2
                    CYCLE
                END IF

                SAME_DIGITS = .TRUE.

                IF (DIGITS_SET_1 /= DIGITS_SET_2) THEN
                    SAME_DIGITS = .FALSE.
                    EXIT
                END IF

            END DO

            IF (SAME_DIGITS) EXIT

            NUM_1X = NUM_1X + 1

        END DO

        IF (SAME_DIGITS) EXIT

        N = N + 1

    END DO

    PRINT ('(I0)'), NUM_1X

END PROGRAM F052
